### PR TITLE
Add ability to show/hide hidden games in all modes

### PR DESCRIPTION
**Pull request by mrnohr**
In Kids/Kiosk mode hidden games are not shown. However, I also want to be able to use Full mode, but still hide those hidden games. This change adds a menu option next to the "favorites only" toggle, and then filters the game list accordingly.

_Note:_ I also removed some trailing whitespace in the files I touched in an attempt to clean up the code a bit.

### CODE REVIEW

**Comment by vlaminck** [comment]
Yes, please. This is a great feature. Cleanup is always nice, too 💯

**Comment by zigurana** [comment]
Hi,
While I am not really opposed to adding this, I am having trouble imagining your use-case here.
Let me explain how I see the Full/Kiosk/Kid modes working:
- **Full**, everything is visible. This mode is used when administrating the system, setting up games / adding metadata etc.
- **Kiosk**, config menu items and otherwise explicitly hidden items from game-lists are removed from view. For most of the time, this is the default mode, as normally it's not much use looking at the config items and retropie menu, cluttering up the interface.
- **Kid**, this is the reduced view mode. This is meant to be the locked-down mode, only listing whitelisted items that you explicitly allow in this view.

So now you want to be able to hide items while in the Full view, and that makes me wonder if there are items you are missing in the Kiosk view that you want to keep while still hiding others...

Can you explain what type of items you want to hide, and why you do not want to switch to Kiosk?

File: es-app/src/guis/GuiGamelistOptions.cpp [hunk 1]
```diff
@@ -89,12 +89,25 @@ GuiGamelistOptions::GuiGamelistOptions(Window* window, SystemData* system) : Gui
...
 				LOG(LogDebug) << "Nothing to show in selected favorites mode, resetting";
-				favorite_only->setState(false);			
+				favorite_only->setState(false);
 				Settings::getInstance()->setBool("FavoritesOnly", favorite_only->getState());
 			}
 		}
 	});
-	
+
+	// Toggle: Show hidden
```
Comment (zigurana) [inline es-app/src/guis/GuiGamelistOptions.cpp:98]: This needs to be in a `if(Settings::getInstance()->getString("UIMode") == "Full")` block, as it does not make sense to show the toggle box in any of the other modes.

File: es-app/src/guis/GuiGamelistOptions.cpp [hunk 2]
```diff
@@ -89,12 +89,25 @@ GuiGamelistOptions::GuiGamelistOptions(Window* window, SystemData* system) : Gui
...
+	show_hidden->setState(Settings::getInstance()->getBool("ShowHidden"));
+	mMenu.addWithLabel("SHOW HIDDEN", show_hidden);
+	addSaveFunc([show_hidden, this] {
+		// First save to settings, in order for filtering to work
+		if (show_hidden->getState() != Settings::getInstance()->getBool("ShowHidden"))
+		{
+			Settings::getInstance()->setBool("ShowHidden", show_hidden->getState());
+			mHiddenStateChanged = true;
+		}
+	});
```
Comment (zigurana) [inline es-app/src/guis/GuiGamelistOptions.cpp]: after this, you'll need to check if there is anything left to show in the view, see above for the same situation in favorites_only.

I agree, that this is likely not an issue with hidden items, as it is with favorites or kid-only stuff, and really, this shouldnt even be handled in this file.
It should be checked with a specific function call to filedata if we are going to encounter this situation more often.
But untill then, it needs to be checked.

**Comment by vlaminck** [comment]
I see your point, but I use this version of EmulationStation exclusively for favorites. I don't use Kiosk mode and I don't use Kids mode. I do, however, alter configs and go into the RetroPie menus a lot. I would absolutely love to be able to hide items in Full mode.

I also think that anyone using the Full mode purely as an admin area would understand hidden vs non-hidden and check the "Show Hidden" option. Plus, if an item was hidden in the first place, they likely marked it as hidden for a reason. I don't think that allowing items to be hidden in Full mode will cause any issues for people using it strictly as an admin area. If that were the case I would argue that the ability to only show favorites in the Full mode should be removed for the same reason (but please don't do that).

**Review by zigurana — CHANGES_REQUESTED** [review]

**Comment by zigurana** [comment]
@vlaminck 
I understand, thanks for explaining. I see no problem with including this, once you've made the changes needed. I'll pick up the larger design issues at some point (basically, creating a new function to check if a new filter has any valid items left to show).

File: es-app/src/guis/GuiGamelistOptions.cpp (hunk 1 above)
Comment (mrnohr) [reply es-app/src/guis/GuiGamelistOptions.cpp:98]: Good point. I'll make the changes

**Comment by mrnohr** [comment]
I made the updates as recommended by @zigurana. Now this option only shows in "Full" mode and if there are only hidden games (which should be unlikely) and you choose to not show them, it will check and revet that setting back to show them.

**Review by zigurana — APPROVED** [review]

**Comment by zigurana** [comment]
And done!

**Comment by zigurana** [comment]
Just want to say, this was my first PR as an owner/maintainer, so thanks!

**Pull request merged**